keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,2246 @@
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+<!--
+**ACTION REQUIRED:** In order to merge code into a release, there must be an
+issue in [kubernetes/enhancements] referencing this KEP and targeting a release
+milestone **before the [Enhancement Freeze](https://git.k8s.io/sig-release/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core
+Kubernetes—i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests for meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Add support to Kubernetes for declaring _quality-of-service_ resources, and
+assigning these to Pods. A quality-of-service (QoS-class) resource is similar
+to other Kubernetes resource types (i.e. native resources such as `cpu` and
+`memory` or extended resources) because you can assign that resource to a
+particular container. However, QoS-class resources are also different from
+those other resources because they are used to assign a _class identifier_,
```

Comment:
Yeah, naming and terminology is hard, I've been struggling with it and I'm not satisfied with it myself. The name QoS-class resources is not good in the sense that this would be a generic mechanism that could be used for other scenarios than pure "QoS", too, as pondered in the "Possible future scenarios" section under "User stories".

I'll sleep over this and try to come up with and suggest something better, more generic. If not for the name at least for better description for this section. Even the word "resource" might be a bit misleading here, maybe it should be _Resource-QoS_, or _QoS-assignments_, or _Class-assignments_ or...

Reply:
> How about "However, QoS-class resources are also different from those other resource because they control access to a shared resource instead of granting exclusive access to a certain amount of capacity." ?

Even this is not necessarily true. It's not necessarily about sharing any resource, but e.g. applying some throttling (or other) parameters. On the plain API level it's sort of tagging the pods (and/or containers), i.e. applying a tag from a set of possible tag-values

Reply:
Prioritized Resource makes sense to me. This sounds a lot like priority and fairness in the apiserver but for individual resources.

Reply:
Thanks @logicalhan for the review. I'm really struggling with the naming bc the K8s API is sort of generic mechanism for "tagging" or assigning pods/containers to classes of different types. Even the term "resource" seems to be misleading into thinking quantitative allocation of resources (like devices or cpus). 

@logicalhan do you have suggestions for the field names in the api etc?

Reply:
Do we have concrete examples of things where we want to apply QoS, but which REALLY don't feel like "resources"?

While "QoS" is somewhat jargon-y, it's pretty well understood in my experience.  Alternately, if this is just about names, we could look for similar words?

"banded resources" ? It sort of implies linearness of bands

"tiered resources" ?  Also implies linearity, which I don't think is necessarily true

"categorized" ?

"qualitative" ?